Metrics collection profiles (#1298)
openshift/enhancements: enhancements/monitoring/scrape-profiles.md

  - @openshift/openshift-team-monitoring
approvers:
  - TBD
api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
  - TBD
creation-date: 2022-12-06
last-updated: yyyy-mm-dd
tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
  - https://issues.redhat.com/browse/MON-2483
see-also:
  - "/enhancements/this-other-neat-thing.md"
replaces:
  - "/enhancements/that-less-than-great-idea.md"
superseded-by:
  - "/enhancements/our-past-effort.md"
---

# Scrape profiles

## Summary

The core OpenShift components ship a large number of metrics. A 4.12-nightly
cluster on AWS currently produces around 350K series by default, and enabling
additional add-ons increases that number. Users have repeatedly asked for a supported
method of making Prometheus consume less memory, either by increasing the scraping

> timeout (which isn't a direction we would like to follow) or by scraping fewer targets 

Author: Good idea, I've added that to the Alternatives

- OCP components that would like to implement scrape profiles would have to provide X monitors. Where X is the number of scrape profiles that we would support.

### Workflow Description

The cluster-monitoring-operator (CMO) will be extended with a new field, `scrapeProfile` under `prometheusK8s`. This new field will then allow users to configure their desired scrape profile while preserving the platform functionality (console, HPA and alerts).

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      scrapeProfile: full 
```

The different profiles would be pre-defined by us. Once a profile is selected CMO then populates the main Prometheus CR to select resources that implement the requested profile -- using [pod|service]MonitorSelector and the label `monitoring.openshift.io/scrape-profile` (profile label) -- this way Prometheus would select monitors from two sets:
- monitors with the profile label and the requested label value (profile)
- monitors without the profile label present (additionally to the current namespace selector).

> This was it would be up to OpenShift teams if they wanted to adopt this feature. Without any change to a monitor, if a user picks a profile in the CMO config, things should work as they did before. When a OpenShift team wants to implement scrape profiles, they need to provide monitors for all profiles making sure they do not provide monitors without the profile label. If a profile label is not used, then a monitor will not be scraped at all for any given profile.

Reviewer: ```suggestion
OpenShift teams can decide if they wanted to adopt this feature. Without any change to a ServiceMonitor, if a user picks a profile in the CMO config, things should work as they did before. When an OpenShift team wants to implement scrape profiles, they need to provide monitors for all profiles, making sure they do not provide monitors without the profile label. If a profile label is not used, then a monitor will not be scraped at all for any given profile.
```

In the beginning the goal is to support 2 profiles:

- `full` (same as today)

> - `operational` (only collect metrics necessary for alerts, recording rules, telemetry and dashboards)

Reviewer: Any reason not to call this `minimal`? I wonder if users might misinterpret the term `operational` and assume this contains all they deem necessary for operation of the cluster.

Author: No reason really, I've changed it to minimal to see what others think of it

```yaml
---
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    monitoring.openshift.io/scrape-profile: full
  name: foo-full
  namespace: openshift-bar
spec:
  endpoints:
    port: metrics
  selector:
    matchLabels:
      app.kubernetes.io/name: foo
---
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    monitoring.openshift.io/scrape-profile: operational
  name: foo-operational
  namespace: openshift-bar
spec:
  endpoints:
    port: metrics
  selector:
    matchLabels:
      app.kubernetes.io/name: foo
  metricRelabelings:
  - sourceLabels: [__name__]
    action: keep
    regex: "requests_total|requests_failed_total"
 ```

#### Variation [optional]

NA

### API Extensions

- Addition of a new field to the `cluster-monitoring-config` ConfigMap

```go
type PrometheusK8sConfig struct {
  // Defines the scraping profile that will be enforced on the platform
  // prometheus instance. Possible values are full and operational.
  //
  // +kubebuilder:validation:Enum=full;operational
  // +optional
	ScrapeProfile string `json:"scrapeProfile,omitempty"`
}
```

### Implementation Details/Notes/Constraints [optional]

> Each OpenShift team that wants to adopt this feature will be responsible for providing the different monitors. This work is not trivial, dependecies beween operators exist, this makes unclear for developers if it's really fine or not for their operator in the "operational" profile to not expose a given metric. To try and help teams with this problem we would provide a tool that would consume a monitor and generate the "operational" monitor for that operator. However for this, the tool would need to have access to an up to date, instalation of OpenShift in order to query the Prometheus instance running in the cluster.

Reviewer: ```suggestion
Each OpenShift team that wants to adopt this feature will be responsible for providing the different monitors. This work is not trivial. Dependencies between operators and their metrics exist. This makes it difficult for developers to determine whether a given metric must be added to the "operational" profile or not. To aid teams with this problem we will provide a tool that would consume a monitor and generate the "operational" monitor for that operator. However for this, the tool would need to have access to an up to date, installation of OpenShift in order to query the Prometheus instance running in the cluster.
```

Author: Much better! I was struggling with this one 😅

---
title: scrape-profiles
authors:
  - @JoaoBraveCoding
reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
  - @openshift/openshift-team-monitoring
approvers:
  - TBD
api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
  - TBD
creation-date: 2022-12-06
last-updated: yyyy-mm-dd
tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
  - https://issues.redhat.com/browse/MON-2483
see-also:
  - "/enhancements/this-other-neat-thing.md"
replaces:
  - "/enhancements/that-less-than-great-idea.md"
superseded-by:

>   - "/enhancements/our-past-effort.md"

Reviewer 2: (nit) I don't think that we have any existing enhancement proposal related to the topic.

---

# Scrape profiles

## Summary

The core OpenShift components ship a large number of metrics. A 4.12-nightly

> cluster on AWS currently produces around 350K series by default, and enabling

Reviewer 2: ```suggestion
cluster on AWS currently produces around 350,000 unique timeseries by default, and enabling
```

additional add-ons increases that number. Users have repeatedly asked for a supported

> method of making Prometheus consume less memory, either by increasing the scraping

Reviewer 2: ```suggestion
method of making Prometheus consume less memory and CPU, either by increasing the scraping
```

> interval or by scraping fewer targets -- for example, modifying the ServiceMonitors
> to drop metrics undesired to some users.

Reviewer 2: I'd remove the last bit since it doesn't really relate to larger scrape intervals or scraping fewer targets.
```suggestion
interval or by scraping fewer targets.
```

> These modifications are currently not possible, because the service monitors deployed to OCP
> are managed by operators and cannot be modified. This proposal outlines a solution for allowing
> users to set a level of scraping aligned with their needs.

Reviewer 2: ```suggestion
These modifications are currently not possible, because the OpenShift operators manage their service/pod monitors (either directly or via the cluster version operator) and any manual modification would be reverted. This proposal outlines a solution for allowing users to set a level of scraping aligned with their needs.
```

> ## Motivation

Reviewer 2: In this section, you can expand on the research the data you've gathered when prototyping.

OpenShift is an opinionated platform, and the default set of metrics has of course 
been crafted to match what we think the majority of users might need. Nevertheless,
users have repeatedly asked for the ability to reduce the amount of memory consumed by
Prometheus either by lowering the Prometheus scrape intervals or by modifying ServiceMonitors.

Users currently can not control the ServiceMonitors scraped by Prometheus since some of the
metrics collected are essential for other parts of the system to function propperly
(console, HPA and alerts). Users also are not allowed to tune the interval at which Prometheus
scrapes targets as this again can have unforeseen results that can hinder the platform, a very
low cadence may overwhelm the platform Prometheus instance a very high interval may render some
of the default alerts ineffective.

The goal of this proposal is to allow users to pick their desired level of scraping while limiting
the impact this might have on the platform, via resources under the control of the
cluster-monitoring-operator and other platform operators.

### User Stories

> - As an OpenShift user, I want to lower the amount of memory consumed by Prometheus in a supported way, so I can choose between different scrape profiles, e.g `full` or `minimal`.

Reviewer 2: The persona would rather be the OpenShift cluster administrator. Also "resources" rather than "memory" since CPU is also impacted?

- Openshift developers may optionally adhere to this change, developers that do not adhere should not be impacted.

### Non-Goals

- Dynamically adjusting metrics scraped at runtime based on heuristics.

## Proposal

This enhancement leverages label selectors to allow cluster-monitoring-operator to select the appropriate monitors ([Pod|Service]Monitor) for a given scrape profile. 

Given this, the following steps would be necessary to implement this enhancement:

- cluster-monitoring-operator has to be updated to support the logic of the new field;
- OCP components that would like to implement scrape profiles would have to provide X monitors. Where X is the number of scrape profiles that we would support.

### Workflow Description

The cluster-monitoring-operator (CMO) will be extended with a new field, `scrapeProfile` under `prometheusK8s`. This new field will then allow users to configure their desired scrape profile while preserving the platform functionality (console, HPA and alerts).

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      scrapeProfile: full 
```

> The different profiles would be pre-defined by us. Once a profile is selected CMO then populates the main Prometheus CR to select resources that implement the requested profile -- using [pod|service]MonitorSelector and the label `monitoring.openshift.io/scrape-profile` (profile label) -- this way Prometheus would select monitors from two sets:

Reviewer 2: we need to define "us", I assume that you mean the monitoring team.

In the beginning the goal is to support 2 profiles:

- `full` (same as today)
- `minimal` (only collect metrics necessary for alerts, recording rules, telemetry and dashboards)

When the cluster admin enables the `minimal` profile, the k8s Prometheus resource would be configured accordingly:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: k8s
  namespace: openshift-monitoring
spec:
  serviceMonitorSelector:
    matchExpressions:
    - key: monitoring.openshift.io/scrape-profile
      operator: NotIn
      values:
      - "full"
```
 
An OpenShift team that would want to support the scrape profiles feature would need to provide 2 monitors for each profile (in this example 1 service monitor per profile).

> ```yaml

Reviewer 2: maybe best to include a real example? The `telemeter-client` service monitor could be a good choice as the minimal profile should select only 2 metric names.

---
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    monitoring.openshift.io/scrape-profile: full
  name: foo-full
  namespace: openshift-bar
spec:
  endpoints:
    port: metrics
  selector:
    matchLabels:
      app.kubernetes.io/name: foo
---
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    monitoring.openshift.io/scrape-profile: minimal
  name: foo-minimal
  namespace: openshift-bar
spec:
  endpoints:
    port: metrics
  selector:
    matchLabels:
      app.kubernetes.io/name: foo
  metricRelabelings:
  - sourceLabels: [__name__]
    action: keep
    regex: "requests_total|requests_failed_total"
 ```

#### Variation [optional]

NA

### API Extensions

- Addition of a new field to the `cluster-monitoring-config` ConfigMap

```go
type PrometheusK8sConfig struct {
  // Defines the scraping profile that will be enforced on the platform
  // prometheus instance. Possible values are full and minimal.
  //
  // +kubebuilder:validation:Enum=full;minimal

>   // +optional

Reviewer 2: the kubebuilder annotations are a bit misleading since the config isn't done via CRD.

### Risks and Mitigations

- How are monitors supposed to be kept up to date? In 4.12 a metric that wasn't being used in an alert is now required, how does the monitor responsible for that metric gets updated?
  - The tool we provide would run in CI and ensure are ServiceMonitors are up to date;

- A new profile is added, what will happen to operators that had implemented scrape profiles but did not implement the latest profile.
  - TBD. Currently, according to the description above, they would not be scraped when the new profile would be picked.

### Drawbacks

- Extra CI cycles

## Design Details

### Open Questions [optional]

- Should we add future profiles? 

> - Should developers have to comply with all profiles?

Reviewer 2: I would add another question about ensuring that all metrics used by the console are still present.

### Test Plan

- Unit tests in CMO to validate that the correct monitors are being selected
- E2E tests in CMO to validate that everything works correctly

> Unsure on this one... (- Testing in openshift/CI to validate that every metrics being used exists in the cluster?)

Reviewer 2: I think that we need to explore this space. We already had the idea to detect alerting rules using metrics that don't exist (see https://github.com/openshift/origin/pull/26787).

### Graduation Criteria

Plan to release as Dev Preview initially.

#### Dev Preview -> Tech Preview

- Ability to utilize the enhancement end to end
- End user documentation
- Sufficient test coverage
- Gather feedback from users rather than just developers

#### Tech Preview -> GA

- More testing (upgrade, downgrade, scale)
- Sufficient time for feedback
- Available by default
- Backhaul SLI telemetry
- Document SLOs for the component
- Conduct load testing
- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)

**For non-optional features moving to GA, the graduation criteria must include
end to end tests.**

#### Removing a deprecated feature

- Announce deprecation and support policy of the existing feature
- Deprecate the feature

### Upgrade / Downgrade Strategy

> - Upgrade and downgrade are not expected to present a significant challenge. The new field is a small layers over existing stable APIs.

Reviewer 2: At least, we need to think about the 4.12 > 4.13 upgrade path. Some of the operators may be upgraded before CMO and there may be a short period of time when the Prometheus operator would reconcile all monitoring objects irrespective of their scrape profile.

### Version Skew Strategy

TBD but I don't think it applies here

### minimal Aspects of API Extensions

TBD but I don't think it applies here

#### Failure Modes

TBD but I don't think it applies here

#### Support Procedures

TBD but I don't think it applies here

## Implementation History

Initial proofs-of-concept:

- https://github.com/openshift/cluster-monitoring-operator/pull/1785

> ## Alternatives

Reviewer 2: another option would be CMO injecting metric relabelings for all service monitors based on the rules being deployed (someone [asked](https://github.com/prometheus-operator/prometheus-operator/pull/4554) for some similar recently in upstream prometheus operator). But 1) CMO acting behind the back of other operators isn't probably a very good idea (less explicit, more brittle) and 2) it's likely to be less efficient because the relabeling configs would be very complex and expensive in terms of processing.

> - Let users configure themselves the Prometheus scraping interval. This solution was discarded since some users might not be fully aware of the impact that changing this interval migh have. Too low of an interval and users might overwelm Prometheus and exaust it's memory. In contrast, too long of an interval might render the default alerts ineffective.

Reviewer 2: we need more details on why long scrape intervals can break alerts.

      certFile: /etc/prometheus/secrets/metrics-client-certs/tls.crt
      keyFile: /etc/prometheus/secrets/metrics-client-certs/tls.key
      serverName: server-name-replaced-at-runtime
  jobLabel: k8s-app
  selector:
    matchLabels:
      k8s-app: telemeter-client
---
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    k8s-app: telemeter-client
    monitoring.openshift.io/scrape-profile: minimal
  name: telemeter-client
  namespace: openshift-monitoring
spec:
  endpoints:
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    interval: 30s
    port: https
    scheme: https
    tlsConfig:
      caFile: /etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt
      certFile: /etc/prometheus/secrets/metrics-client-certs/tls.crt
      keyFile: /etc/prometheus/secrets/metrics-client-certs/tls.key
      serverName: server-name-replaced-at-runtime
  jobLabel: k8s-app
  selector:
    matchLabels:
      k8s-app: telemeter-client
  metricRelabelings:
  - sourceLabels: [__name__]
    action: keep
    regex: "federate_samples|federate_filtered_samples"

>  ```

Reviewer: Does it make sense to drop the not-so-relevant fields here? E.g `tlsConfig` is not really needed to illustrate the functionality.
Maybe a textual note like `Note the metricRelabelings section only keeping two metrics, while the rest is dropped.` is also useful?

	ScrapeProfile string `json:"scrapeProfile,omitempty"`
}
```

### Implementation Details/Notes/Constraints [optional]

Each OpenShift team that wants to adopt this feature will be responsible for providing the different monitors. This work is not trivial. Dependencies between operators and their metrics exist. This makes it difficult for developers to determine whether a given metric must be added to the "minimal" profile or not. To aid teams with this problem we will provide a tool that would consume a monitor and generate the "minimal" monitor for that operator. However for this, the tool would need to have access to an up to date, installation of OpenShift in order to query the Prometheus instance running in the cluster.

### Risks and Mitigations

- How are monitors supposed to be kept up to date? In 4.12 a metric that wasn't being used in an alert is now required, how does the monitor responsible for that metric gets updated?
  - The tool we provide would run in CI and ensure are ServiceMonitors are up to date;

- A new profile is added, what will happen to operators that had implemented scrape profiles but did not implement the latest profile.

>   - TBD. Currently, according to the description above, they would not be scraped when the new profile would be picked.

Author: > It'll probably be up the the CMO dev team to either chase down the ServiceMonitor owners or provide a SM for the new profile. 

I wouldn't want to go down this path, simply because this way we would have to track which components were using the scrape profile feature. As an example currently, I don't think we would be able to come up with a list of components that currently deploying a ServiceMonitor to be consumed by `openshift-monitoring`. Unless I'm missing some info regarding this 👀 

> A safe default would be to simply copy the full service monitor for any given new profile.

I like this but we would have to come up with a strategy on how this would be implemented 🤔 because selectors just give us a set of resources, so optional conditions are hard

>   - TBD. Currently, according to the description above, they would not be scraped when the new profile would be picked.

Author: After a chat with @simonpasquier I've added some lines regarding this issue, this section should be reworked

of the default alerts ineffective.

The goal of this proposal is to allow users to pick their desired level of scraping while limiting
the impact this might have on the platform, via resources under the control of the
cluster-monitoring-operator and other platform operators.

Furthermore to assess the viability of the scrape profile feature the monitoring
team performed a detailed analysis of its impact in an OpenShift cluster.
The analysis consisted in a test where an OpenShift cluster would run three replicas
of the OpenShift Prometheus instance but each replica would be configured to a 
different scrape profile (`full`, `minimal`, `uponly`). Then we would trigger a 
workload using kube-burner and at the end of 2 hours, we evaluated the results.
We concluded that in terms of resource usage, given the results obtained we
can confidently state that the feature is quite valuable given the reduction of
CPU by ~21% and memory by ~33% when comparing the `minimal` to the `full` profile.
More detail can be consulted in this [document](https://docs.google.com/document/d/1MA-HTJQ_X7y_bwpJS2IPbGmC4qDyIMp25jisr34X2F4/edit?usp=sharing)

> 

Reviewer 2: I've found additional data for this section based on some Telemetry numbers: since a few releases, we collect for each cluster the top-3 Prometheus jobs generating most samples.
Based on the 4.11 numbers, the 5 components most often reported are: the Kubernetes API servers, the Kubernetes schedulers, kube-state-metrics, kubelet and the networ daemon.

Author: That's neat! Can you share the query that gave you that info?

### User Stories

- As an OpenShift cluster administrator, I want to lower the amount of resources consumed by Prometheus in a supported way, so I can choose between different scrape profiles, e.g `full` or `minimal`.

> - As an OpenShift developer, I want a supported way to collect a subset of the metrics exported by my operator depending on the distribution.

Reviewer 2: ```suggestion
- As an OpenShift developer, I want a supported way to collect a subset of the metrics exported by my operator and operands.
```

### Goals

- Give users a supported and documented method to pick the amount of metrics being consumed by the platform, in a way that allows cluster-monitoring-operator and other operators to provide their own defaulting.
- Other components of the cluster (console, HPA, alerts) should not be negatively affected by this change.
- Openshift developers may optionally adhere to this change, developers that do not adhere should not be impacted.

### Non-Goals

- Dynamically adjusting metrics scraped at runtime based on heuristics.
- Allowing users to explicitly set a list or a regex of metrics that they would
want to be included in a given profile.

## Proposal

This enhancement leverages label selectors to allow cluster-monitoring-operator to select the appropriate monitors ([Pod|Service]Monitor) for a given scrape profile. 

Given this, the following steps would be necessary to implement this enhancement:

- cluster-monitoring-operator has to be updated to support the logic of the new field;
- OCP components that would like to implement scrape profiles would have to provide X monitors. Where X is the number of scrape profiles that we would support.

> ### Workflow Description

Reviewer 2: we need to explain what happens if the cluster picks an unsupported scrape profile value (e.g. CMO should probably fail the reconciliation and report a Degraded=True condition).

Author: So, that's an interesting point which I hadn't thought about. CMO in on itself (in the current implementation) doesn't run any validation of the scrape profiles that exist. So currently is not able to report a `Degraded=True` condition.
However, this might be an interesting idea to explore as it might also solve https://github.com/openshift/enhancements/pull/1298#discussion_r1067067971 what if CMO would periodically pull all monitors, and then it would:
1. Go through all monitors and extract the services/pods
2. For each service it would either have to find one or len(scrapeProfiles) monitor(s)
3. Either report `Degraded=Flase` or `Degraded=True` if the above check was not successful

Reviewer 2: :+1:

The cluster-monitoring-operator (CMO) will be extended with a new field, `scrapeProfile` under `prometheusK8s`. This new field will then allow users to configure their desired scrape profile while preserving the platform functionality (console, HPA and alerts).

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      scrapeProfile: full 
```

> The different profiles would be pre-defined by the OpenShift monitoring team. Once a profile is selected CMO then populates the main Prometheus CR to select resources that implement the requested profile -- using [pod|service]MonitorSelector and the label `monitoring.openshift.io/scrape-profile` (profile label) -- this way Prometheus would select monitors from two sets:

Reviewer 2: ```suggestion
The different profiles would be pre-defined by the OpenShift monitoring team. Once a profile is selected CMO then updates the platform Prometheus CR to select pod and service monitors as the union of the 2 sets:
```

- monitors with the profile label and the requested label value (profile)
- monitors without the profile label present (additionally to the current namespace selector).

> OpenShift teams can decide if they wanted to adopt this feature. Without any change to a ServiceMonitor, if a user picks a profile in the CMO config, things should work as they did before. When an OpenShift team wants to implement scrape profiles, they need to provide monitors for all profiles, making sure they do not provide monitors without the profile label. If a profile label is not used, then a monitor will not be scraped at all for any given profile.

Author: This slipped through the cracks, it doesn't make sense. What I should write here is, the following

> If a team implements scrape profiles they must ensure that all their monitors have the label `monitoring.openshift.io/scrape-profile` set with the appropriate value, otherwise their component might end up being double scraped or not scraped at all.

I added it

In the beginning the goal is to support 2 profiles:

- `full` (same as today)
- `minimal` (only collect metrics necessary for alerts, recording rules, telemetry and dashboards)

When the cluster admin enables the `minimal` profile, the k8s Prometheus resource would be configured accordingly:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: k8s
  namespace: openshift-monitoring
spec:
  serviceMonitorSelector:
    matchExpressions:
    - key: monitoring.openshift.io/scrape-profile
      operator: NotIn
      values:
      - "full"
```
 

> An OpenShift team that would want to support the scrape profiles feature would need to provide 2 monitors for each profile (in this example 1 service monitor per profile).

Reviewer 2: ```suggestion
An OpenShift team that wants to support the scrape profiles feature would need to provide 2 monitors for each profile (in this example 1 service monitor per profile).
```

kind: ServiceMonitor
metadata:
  labels:
    k8s-app: telemeter-client
    monitoring.openshift.io/scrape-profile: minimal
  name: telemeter-client
  namespace: openshift-monitoring
spec:
  endpoints:
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    interval: 30s
    port: https
    scheme: https
    tlsConfig:
      <...>
  jobLabel: k8s-app
  selector:
    matchLabels:
      k8s-app: telemeter-client
  metricRelabelings:
  - sourceLabels: [__name__]
    action: keep
    regex: "federate_samples|federate_filtered_samples"
 ```

Note: the `metricRelabeling` section only keeping two metrics, while the rest is dropped.

Finally, a team that adopts the scrape profile feature should also add 
themselves and their component to the list under 
[Infrastruture needed](#infrastructure-needed-optional)

#### Variation [optional]

NA

### API Extensions

- Addition of a new field to the `cluster-monitoring-config` ConfigMap

```go
type PrometheusK8sConfig struct {
  // Defines the scraping profile that will be enforced on the platform
  // prometheus instance. Possible values are full and minimal.
	ScrapeProfile string `json:"scrapeProfile,omitempty"`
}
```

> ### Implementation Details/Notes/Constraints [optional]

Reviewer 2: ```suggestion
### Implementation Details/Notes/Constraints

### Risks and Mitigations

- How are monitors supposed to be kept up to date? In 4.12 a metric that wasn't being used in an alert is now required, how does the monitor responsible for that metric gets updated?
  - The tool we provide would run in CI and ensure are ServiceMonitors are up to date;

- A new profile is added, what will happen to operators that had implemented scrape profiles but did not implement the latest profile.
  - The monitoring team will use the list under 
  [Infrastruture needed](#infrastructure-needed-optional) to help the developers
  with the addoption of the new scrape profile.

### Drawbacks

- Extra CI cycles

## Design Details

### Open Questions [optional]

- Should we add future profiles? 

> - Should developers have to comply with all profiles?

Reviewer 2: I guess that this question is already answered above as "yes, you need to implement all profiles or nothing".

- How will ensure that all metrics used by the console are still present?

### Test Plan

- Unit tests in CMO to validate that the correct monitors are being selected
- E2E tests in CMO to validate that everything works correctly
Unsure on this one... (- Testing in openshift/CI to validate that every metrics being used exists in the cluster?)

### Graduation Criteria

> Plan to release as Dev Preview initially.

Reviewer 2: we could target TechPreview as the first step: the default being "full", it shouldn't affect operations.

#### Dev Preview -> Tech Preview

- Ability to utilize the enhancement end to end
- End user documentation
- Sufficient test coverage
- Gather feedback from users rather than just developers

#### Tech Preview -> GA

- More testing (upgrade, downgrade, scale)
- Sufficient time for feedback
- Available by default
- Backhaul SLI telemetry
- Document SLOs for the component
- Conduct load testing
- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)

**For non-optional features moving to GA, the graduation criteria must include
end to end tests.**

#### Removing a deprecated feature

- Announce deprecation and support policy of the existing feature
- Deprecate the feature

### Upgrade / Downgrade Strategy

> - If this feature is accepted and is released to 4.13 then we should backport it as well to 4.12. The reason is that when the upgrade occurs and the new ServiceMonitors are deployed to the cluster, we don't want the 4.12 Prometheus-Operator to configure the 4.12 Prometheus instances to start double scraping OpenShift components that deployed the new ServiceMonitors.

Reviewer 2: "feature" needs to be defined better. I suppose that it means updating the pod & service monitor selectors for the k8s prometheus.

  - This container would perform an analysis on what metrics were being used. Then it would provide prometheus operator with this list.
  - Prometheus-operator with the list provided by this new component would know what scraping targets it could change to keep certain metrics.
- Recently Azure also added support for scrape profiles:
  - [Azure Docs](https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-metrics-scrape-configuration-minimal)
  -  https://github.com/Azure/prometheus-collector
  - In their approach they also have [hard coded](https://github.com/Azure/prometheus-collector/blob/66ed1a5a27781d7e7e3bb1771b11f1da25ffa79c/otelcollector/configmapparser/tomlparser-default-targets-metrics-keep-list.rb#L28) 
  set of metrics that are only consussumed when the minimal profile is enabled.
  However, customer are also able to extend this minimal profile with regexes
  to include metrics which might be interesting to them.

## Infrastructure Needed [optional]

### Adopted Scrape Profiles

Add the team and the component that wants to adopt scrape profiles

```yaml
----- Team -----| ----- Component -----

> Monitoring Team | Cluster Monitoring Operator

Reviewer 2: We can even detail by component, e.g. kube-state-metrics and kubelet monitors are the main candidates for us. There should also be a "completion status" or something similar to differentiate between "considering", "implementation in progress" and "implemented"

for OpenShift 4.11 the 5 components most often reported as the biggest producers
are: the Kubernetes API servers, the Kubernetes schedulers, kube-state-metrics,
kubelet and the network daemon.

### User Stories

- As an OpenShift cluster administrator, I want to lower the amount of resources
  consumed by Prometheus in a supported way, so I can choose between different
  scrape profiles, e.g `full` or `minimal`.
- As an OpenShift developer, I want a supported way to collect a subset of the
  metrics exported by my operator and operands.

### Goals

- Give users a supported and documented method to pick the amount of metrics
  being collected by the platform, in a way that allows
  cluster-monitoring-operator and other operators to provide their own
  defaulting.
- Other components of the cluster (console, HPA, alerts) should not be
  negatively affected by this change.
- Openshift developers may optionally adhere to this change, developers that do
  not adhere should not be impacted.

### Non-Goals

- Dynamically adjusting metrics scraped at runtime based on heuristics.
- Allowing users to explicitly set a list or a regex of metrics that they would

> want to be included in a given profile.

Reviewer 3: IIUC the second case [here](https://issues.redhat.com/browse/MON-2694?focusedId=21834890&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-21834890) suggests otherwise?

Author: Here with `user` I mean OpenShift end users while in the link you provided I mean OpenShift developer

metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      scrapeProfile: full 
```

The different profiles would be pre-defined by the OpenShift monitoring team.
Once a profile is selected CMO then updates the platform Prometheus CR to select
pod and service monitors as the union of the 2 sets:
- pod and service monitors with the 
  `monitoring.openshift.io/scrape-profile: <selected profile>` label.
- pod and service monitors without the 
  `monitoring.openshift.io/scrape-profile` profile label present.

OpenShift teams can decide if they wanted to adopt this feature. Without any
change to a ServiceMonitor, if a user picks a profile in the CMO config, things
should work as they did before. When an OpenShift team wants to implement scrape
profiles, they need to provide monitors for all profiles, making sure they do
not provide monitors without the profile label. If a team implements scrape
profiles they must ensure that all their monitors have the label
`monitoring.openshift.io/scrape-profile` set with the appropriate value,
otherwise their component might end up being double scraped or not scraped at

> all.

Author: With the changes I did, I think this is clearer now, do let me know

  metricRelabelings:
  - sourceLabels: [__name__]
    action: keep
    regex: "federate_samples|federate_filtered_samples"
 ```

Note: the `metricRelabeling` section only keeping two metrics, while the rest is
dropped.

Finally, a team that adopts the scrape profile feature should also add 
themselves and their component to the list under 
[Infrastruture needed](#infrastructure-needed-optional)

#### Variation [optional]

NA

### API Extensions

- Addition of a new field to the `cluster-monitoring-config` ConfigMap

```go
type PrometheusK8sConfig struct {
  // Defines the scraping profile that will be enforced on the platform
  // prometheus instance. Possible values are full and minimal.
	ScrapeProfile string `json:"scrapeProfile,omitempty"`
}
```

### Implementation Details/Notes/Constraints

Each OpenShift team that wants to adopt this feature will be responsible for
providing the different monitors. This work is not trivial. Dependencies between
operators and their metrics exist. This makes it difficult for developers to
determine whether a given metric must be added to the "minimal" profile or not.
To aid teams with this problem we will provide a tool that would consume a

> monitor and generate the "minimal" monitor for that operator. However for this,

Author: Yes, I've re-written this block to more clearly show that

the tool would need to have access to an up to date, installation of OpenShift
in order to query the Prometheus instance running in the cluster.

### Risks and Mitigations

- How are monitors supposed to be kept up to date? In 4.12 a metric that wasn't

>   being used in an alert is now required, how does the monitor responsible for

Author: As discussed this will be done by a CI test that will validate that all resources (Alerts, PrometheusRule and Dashboards) are metrics that are present in the ServiceMonitors

  that metric gets updated?
  - The tool we provide would run in CI and ensure are ServiceMonitors are up to

>     date;

Author: Re did this section

- A new profile is added, what will happen to operators that had implemented
  scrape profiles but did not implement the latest profile.
  - The monitoring team will use the list under 
  [Infrastruture needed](#infrastructure-needed-optional) to help the developers
  with the addoption of the new scrape profile.

### Drawbacks

> - Extra CI cycles

Reviewer 3: Edit: This is not the case anymore, we'll move forward with this solution in tooling instead (as per the discussion).

## Design Details

### Open Questions

- Should we add future profiles? 
- How will ensure that all metrics used by the console are still present?
- What happens if CMO pick-up an unsupported scrape value? (dicussion on
  https://github.com/openshift/enhancements/pull/1298#discussion_r1072266853)

### Test Plan

- Unit tests in CMO to validate that the correct monitors are being selected

> - E2E tests in CMO to validate that everything works correctly Unsure on this
> one... (- Testing in openshift/CI to validate that every metrics being used
> exists in the cluster?)

Reviewer 3: ```suggestion
- E2E tests in CMO to validate that everything works correctly.
```
> [...] every metrics being used exists in the cluster?

Expanding on the comment above, there can two cases to this. For internal metrics, we can maintain a list of all required metrics, whereas for external metrics, these can be pushed to CMO by the respective team (single source of truth). The operator will validate them all the same.

Author: Same as https://github.com/openshift/enhancements/pull/1298#discussion_r1267919544

### Graduation Criteria

Plan to release as TechPreview as the first step: the default being `full`, it shouldn't impact operations.

#### Tech Preview -> GA

- More testing (upgrade, downgrade, scale)
- Sufficient time for feedback
- Available by default
- Backhaul SLI telemetry
- Document SLOs for the component
- Conduct load testing

> - User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)

Reviewer 3: This will need to be updated based on [MON-3043](https://issues.redhat.com/browse/MON-3043).

    Staleness occurs at 5 minutes (by default) which is would cause gaps in
    graphs/dashboards;
  - Many upstream dashboards were built while having the 30 second interval in
    mind;
  - Scrapes can fail, but the user might not be mindful of this and set a high
    scrape interval;
- Add a seperate container to prometheus-operator (p-o) that would be used by
  p-o to modify the prometheus config according to a scrape profile.
  - This container would perform an analysis on what metrics were being used.
    Then it would provide prometheus operator with this list.
  - Prometheus-operator with the list provided by this new component would know
    what scraping targets it could change to keep certain metrics.
- Recently Azure also added support for scrape profiles:
  - [Azure
    Docs](https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-metrics-scrape-configuration-minimal)
  -  https://github.com/Azure/prometheus-collector
  - In their approach they also have [hardcoded](https://github.com/Azure/prometheus-collector/blob/66ed1a5a27781d7e7e3bb1771b11f1da25ffa79c/otelcollector/configmapparser/tomlparser-default-targets-metrics-keep-list.rb#L28)

>   set of metrics that are only consussumed when the minimal profile is enabled.

Reviewer 3: ```suggestion
  set of metrics that are only consumed when the minimal profile is enabled.
```

  However, customer are also able to extend this minimal profile with regexes to

>   include metrics which might be interesting to them.

Reviewer 3: It's worth considering if we want to allow customers to exercise this functionality, while subjecting the operator to the risk of high memory consumption even for minimal profiles.

Author: Indeed, however for now that's a non-goal and it's out of the scope, it's listed here so reviewers are aware what other products support

## Infrastructure Needed [optional]

### Adopted Scrape Profiles

Add the team and the component that will to adopt scrape profiles and
implementation status.
Possible implementation status: 
- considering
- implementation in progress
- implemented

| Team            | Component          | Implementation Status      |
|-----------------|--------------------|----------------------------|
| Monitoring Team | kubelet            | Implementation in progress |
| Monitoring Team | etcd               | Implementation in progress |
| Monitoring Team | kube-state-metrics | Implementation in progress |
| Monitoring Team | node-exporter      | Implementation in progress |

> | Monitoring Team | prometheus-adapter | Implementation in progress |

Reviewer 3: These are all implemented as a part of https://github.com/openshift/cluster-monitoring-operator/pull/1785 IIUC.
